addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go

```diff
@@ -171,15 +209,203 @@ func addMissingFields(copyFrom, destination map[string]interface{}) error {
+			for _, cbPackage := range packages {
+				if cbPackage != nil && cbPackage.ValuesFrom == nil {
+					nilValuesFromPackages = append(nilValuesFromPackages, cbPackage)
+				}
+			}
+
+			if err := h.AddMissingSpecFieldsFromTemplate(clusterBootstrapTemplate, clusterBootstrap, nil); err != nil {
+				h.Logger.Error(err, fmt.Sprintf("unable to add missing spec fields of ClusterBootstrap %s/%s from ClusterBootstrapTemplate %s/%s",
+					clusterBootstrap.Namespace, clusterBootstrap.Name, clusterBootstrapTemplate.Namespace, clusterBootstrapTemplate.Name))
+				return nil, err
+			}
+
+			// AddMissingSpecFieldsFromTemplate() updates clusterBootstrap.Spec, it holds the updated pointers now.
+			// Declaring updatedCBPackages is to compare with original nilValuesFromPackages and figure out which packages
+			// need to be cloned.
+			updatedCBPackages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+				clusterBootstrap.Spec.CNI,
+				clusterBootstrap.Spec.CPI,
+				clusterBootstrap.Spec.CSI,
+				clusterBootstrap.Spec.Kapp,
+			}, clusterBootstrap.Spec.AdditionalPackages...)
+
+			var packagesToBeCloned []*runtanzuv1alpha3.ClusterBootstrapPackage
+			for _, nilValuesFromPackage := range nilValuesFromPackages {
+				for _, updatedCBPackage := range updatedCBPackages {
+					// If the nilValuesFromPackage has been filled by AddMissingSpecFieldsFromTemplate(), it is the package
+					// we want to record and do cloning.
+					if updatedCBPackage != nil && updatedCBPackage.ValuesFrom != nil && updatedCBPackage.RefName == nilValuesFromPackage.RefName {
+						packagesToBeCloned = append(packagesToBeCloned, updatedCBPackage)
+					}
+				}
+			}
+			packages = packagesToBeCloned
+		}
+	}
+
+	secrets, providers, err := h.CloneReferencedObjectsFromCBPackages(cluster, packages, systemNamespace)
+	if err != nil {
+		h.Logger.Error(err, "unable to clone secrets or providers")
+		return nil, err
+	}
+
+	clusterBootstrap.OwnerReferences = []metav1.OwnerReference{
+		{
+			APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+			Kind:               cluster.Kind,
+			Name:               cluster.Name,
+			UID:                cluster.UID,
+			Controller:         pointer.BoolPtr(true),
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		},
+	}
+	if err := h.K8sClient.Update(h.Ctx, clusterBootstrap); err != nil {
+		return nil, err
+	}
+	clusterBootstrap.Status.ResolvedTKR = tkrName
+	if err := h.K8sClient.Status().Update(h.Ctx, clusterBootstrap); err != nil {
+		return nil, err
+	}
+
+	// ensure ownerRef of clusterBootstrap on created secrets and providers, this can only be done after
+	// clusterBootstrap is created
+	if err := h.EnsureOwnerRef(clusterBootstrap, secrets, providers); err != nil {
+		h.Logger.Error(err, fmt.Sprintf("unable to ensure ClusterBootstrap %s/%s as a ownerRef on created secrets and providers", clusterBootstrap.Namespace, clusterBootstrap.Name))
+		return nil, err
+	}
```

Comment:
Question: Should the annotation (clusterBootstrap.Annotations[constants.AddCBMissingFieldsAnnotationKey]) be removed as well?